File: 155. Min Stack/MinStack.java
class MinStack {
    Stack<Integer> stk = new Stack<Integer>();
    Stack<Integer> minValue = new Stack<Integer>();
    /** initialize your data structure here. */
    public MinStack() {
        
    }
    
    public void push(int x) {
        stk.push(x);
        if (minValue.empty()) {
            minValue.push(x);
        }
        else {
            if (x <= minValue.peek()) {
                minValue.push(x);
            }
        }
    }
    
    public void pop() {
        if (!stk.empty()) {
            int temp = stk.pop();
            if (temp == minValue.peek()) {
                minValue.pop();
            }
        }
    }
    
    public int top() {
        return stk.peek();
    }
    
    public int getMin() {
        return minValue.peek();
    }
}

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack obj = new MinStack();
 * obj.push(x);
 * obj.pop();
 * int param_3 = obj.top();
 * int param_4 = obj.getMin();
 */

File: 155. Min Stack/MinStack.py
class MinStack:
    def __init__(self):
        """
        initialize your data structure here.
        """
        self.stack = []
        self.minValue = []

    def push(self, x: int) -> None:
        self.stack.append(x)
        if not self.minValue:
            self.minValue.append(x)
        else:
            if x <= self.minValue[-1]:
                self.minValue.append(x)

    def pop(self) -> None:
        if self.stack:
            temp = self.stack.pop()
            if temp == self.minValue[-1]:
                self.minValue.pop()

    def top(self) -> int:
        return self.stack[-1] if len(self.stack) else None

    def getMin(self) -> int:
        return self.minValue[-1] if len(self.minValue) else None



# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(x)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()